### PR TITLE
Helm: add ruler specific service account

### PR DESCRIPTION
**Pull request by QuantumEnigmaa**
#### What this PR does

This PR adds a specific service account for the `ruler` component so that different permissions/authorizations can be given to the ruler. 
This can help when using IRSA for example.

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by QuantumEnigmaa** [comment]
If maintainers like the idea, I could propagate this change to all components

**Comment by dimitarvdimitrov** [comment]
the idea makes sense to me. I can't think of a technical reason to have a single service account for all pods. @krajorama WDYT?

**Comment by QuantumEnigmaa** [comment]
Hey @dimitarvdimitrov any news on this one ?

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/_helpers.tpl:81]: shouldn't this use the `.Values.ruler.serviceAccount.create` instead of `.Values.serviceAccount.create`?

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/_helpers.tpl:85]: let's just fall back to the `"mimir.serviceAccountName"` without adding additional default logic. So something like this (haven't tested it)
```suggestion
    {{ include "mimir.serviceAccountName" . }}
```

File: operations/helm/charts/mimir-distributed/templates/ruler/ruler-sa.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/ruler/ruler-sa.yaml:5]: should this use `mimir.ruler.serviceAccountName`?

File: operations/helm/charts/mimir-distributed/CHANGELOG.md
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/CHANGELOG.md:46]: can you move this one line up, add the number of the PR at the end, and also add the actual value that needs to be set `ruler.serivceAcount.create`?

File: operations/helm/charts/mimir-distributed/values.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/values.yaml:1126]: can you add a comment explaining how this relates to the default service account? i.e. the general service account can be disabled and this can still be enabled

**Review by dimitarvdimitrov — COMMENTED** [review]
> If maintainers like the idea, I could propagate this change to all components


Is this necessary on all components? Is this necessary in your use case? You mentioned IRSA - can you explain what that is and why a separate SA is necessary?

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl
Comment (QuantumEnigmaa) [reply operations/helm/charts/mimir-distributed/templates/_helpers.tpl:81]: yeah you're right. Let me fix that.

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl
Comment (QuantumEnigmaa) [reply operations/helm/charts/mimir-distributed/templates/_helpers.tpl:85]: Fine with me.

File: operations/helm/charts/mimir-distributed/templates/ruler/ruler-sa.yaml
Comment (QuantumEnigmaa) [reply operations/helm/charts/mimir-distributed/templates/ruler/ruler-sa.yaml:5]: Absolutely, otherwise, it wouldn't make much sense :facepalm:

File: operations/helm/charts/mimir-distributed/CHANGELOG.md
Comment (QuantumEnigmaa) [reply operations/helm/charts/mimir-distributed/CHANGELOG.md:46]: Sure.

File: operations/helm/charts/mimir-distributed/values.yaml
Comment (QuantumEnigmaa) [reply operations/helm/charts/mimir-distributed/values.yaml:1126]: Sure.

**Comment by QuantumEnigmaa** [comment]
> Is this necessary on all components? 

Not really, at least not from my use case perspective for now. But having the possibility to fine tune each and every component is always better than having a monolithic, unified service account for all.

> Is this necessary in your use case? You mentioned IRSA - can you explain what that is and why a separate SA is necessary?

Yes, in my use case this is a must. [IRSA](https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/setting-up-enable-IAM.html) is short for **Iam Role for Service Account**, It is a way to access resources from AWS from an AWS Kubernetes cluster by having service accounts with a specific annotation that grant a IAM Role to the pods using it. 

In our case, we use this to have mimir pods accessing S3 buckets. We want to have a separate bucket for the `ruler` along with a separate IAM Role for accessing it, different from the one used by all other mimir components. And since for this we need a separate serviceaccount with a different annotation, that's why we need this PR to be merged

**Comment by QuantumEnigmaa** [comment]
@dimitarvdimitrov I disabled the ruler specific sa by default. I also added some lines in the `rolebinding` template so that the ruler service account is still linked to the same role as other mimir pods

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/_helpers.tpl:82]: should be this be `.Values.ruler.serviceAccount.name` instead of `.Values.serviceAccount.name`?

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/_helpers.tpl:88]: there are two approaches to indentation within this function. Let's use just one.

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/_helpers.tpl:83]: so if I provide

```yaml
ruler:
  serviceAccont: 
    name: ruler-sa
```

the service account will have the name `ruler-sa-ruler`. Is the `-ruler` suffix necessary here?

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/_helpers.tpl:81]: the behaviour with the default SA is that people might opt to not have the chart create it, but still have the pods associate with the name there. Should we take the same approach? 

Instead of `if .Values.ruler.serviceAccount.create` we can do `if .Values.ruler.serviceAccount.name`. The default of `null` for the name in values.yaml will just end up using general SA.

I know I'm suggesting to change this line for the second time, sorry about that

**Review by dimitarvdimitrov — COMMENTED** [review]
thanks for addressing my previous comments.

I also see that the linter found some trailing spaces in values.yaml (visible in the diff view of the PR and in the failing CI action). Can you fix those too, please?

File: operations/helm/charts/mimir-distributed/templates/_helpers.tpl
Comment (QuantumEnigmaa) [reply operations/helm/charts/mimir-distributed/templates/_helpers.tpl]: Done

**Comment by QuantumEnigmaa** [comment]
@dimitarvdimitrov I updated the PR by taking into account your comments but I did it in a "personal" way, or so-to-say. 
I kept the `.Values.ruler.serviceAccount.create` field but I gave 3 possibilities for the user : 

* The user do not need a specific sa for the `ruler` so he doesn't change the default values for the `Values.ruler.serviceAccount` section. 
  --> This will lead the `ruler` pods to use the default sa as all the other components.
* The user need a specific sa for the `ruler` but doesn't care for its name so he set `.Values.ruler.serviceAccount.create` to `true` but do not change the default value for `.Values.ruler.serviceAccount.name`. 
  --> This will lead the `ruler` pods to use a dedicated sa named after the default sa for mimir with the `-ruler` suffix.
* The user need a specific sa for the `ruler` and want it to have a specific name so he set `.Values.ruler.serviceAccount.create` to `true` and updates the `.Values.ruler.serviceAccount.name` with a specific value.
  --> This will lead the `ruler` pods to use a dedicated sa which name will be the one from the `.Values.ruler.serviceAccount.name` field.

Tell me if that's ok for you. Otherwise I can go with what you proposed, getting rid of the `.Values.ruler.serviceAccount.create` field and only relying on the `.Values.ruler.serviceAccount.name` one to either create the dedicated sa or not. :)

**Comment by QuentinBisson** [comment]
Hello @dimitarvdimitrov would you have some time to help us push this forward this week?

File: operations/helm/charts/mimir-distributed/values.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/values.yaml:1122]: can you add a doc comment (`# --`) about the meaning of the empty string?

**Review by dimitarvdimitrov — COMMENTED** [review]
Amazing! I think your approach looks great. I only left one minor comment, but otherwise LGTM. After you address that we can merge.

_Sorry about the delay, I was away last week and travelling back from FOSDEM yesterday._

File: operations/helm/charts/mimir-distributed/values.yaml
Comment (QuantumEnigmaa) [reply operations/helm/charts/mimir-distributed/values.yaml:1124]: Done :)

**Comment by QuentinBisson** [comment]
@dimitarvdimitrov I am sorry for pushing again as this was blocking us from moving foward with mimir so I would really like to thank you for helping us out here :) 

I'm not blaming you at all on the delays btw (I know what it's like to maintain open source) and I hope FOSDEM was great :)

I am a just a bit puzzled as it seems to me like you are the only chart maintainer of such a big chart and I thought there would be other maintainers :)

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, thanks

I suspect the chart is not the most exciting piece of code in this repo for other maintainers and other maintainers have move on to other projects within Mimir

**Comment by dimitarvdimitrov** [comment]
FYI this change will be included in next week's development release

**Pull request merged**
